controllers/node-delete/controller.go

```
@@ -0,0 +1,116 @@
+//
+// Events are not sent as they require an object. By this point the namespace
+// object will no longer be available.
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	err := r.api.DeleteNode(req.Name)
+	if err != nil && err != storageos.ErrNodeNotFound {
+		// Re-queue without error.  We will get frequent transient errors, such
+		// as version conflicts or locked objects - that's ok.  Even refusal to
+		// delete due to the node still showing as online should be seen as
+		// transient and should not raise an error.
+		return ctrl.Result{Requeue: true}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+// ChangePredicate filters events before enqueuing the keys.
+type ChangePredicate struct {
+	predicate.Funcs
+}
+
+// Create determines whether an object create should trigger a reconcile.
+func (ChangePredicate) Create(event.CreateEvent) bool {
+	return false
+}
+
+// Update determines whether an object update should trigger a reconcile.
+func (ChangePredicate) Update(event.UpdateEvent) bool {
+	return false
+}
+
+// Delete determines whether an object delete should trigger a reconcile.
+func (ChangePredicate) Delete(e event.DeleteEvent) bool {
+	// Ignore objects without metadata.
+	if e.Meta == nil {
+		return false
+	}
+	return hasStorageOSDriverAnnotation(e.Meta.GetAnnotations())
+}
+
+// Generic determines whether an generic event should trigger a reconcile.
+func (ChangePredicate) Generic(event.GenericEvent) bool {
+	return false
+}
+
+// hasStorageOSDriverAnnotation returns true if the node has the StorageOS
+// driver annotation.  The annotation is added by the drivers' node registrar,
+// so should only be present on nodes that have run StorageOS.
+func hasStorageOSDriverAnnotation(annotations map[string]string) bool {
+	drivers, ok := annotations[DriverAnnotationKey]
+	if !ok {
+		return false
+	}
+	driversMap := map[string]string{}
+	if err := json.Unmarshal([]byte(drivers), &driversMap); err != nil {
+		return false
```

Logging this error would be helpful in debugging when it fails. Avoiding silent failures.

---

Agreed, added